src/write.rs

```diff
@@ -149,22 +149,59 @@ struct ZipRawValues {
     compressed_size: u64,
     uncompressed_size: u64,
 }
-
+trait Sealed {}
+/// File options Extensions
+#[allow(private_bounds)]
+#[doc(hidden)]
+pub trait FileOptionExtension: Default + Sealed {
+    /// Extra Data
+    fn extra_data(&self) -> Option<Arc<Vec<u8>>>;
+    /// Central Extra Data
+    fn central_extra_data(&self) -> Option<Arc<Vec<u8>>>;
+}
+impl Sealed for () {}
+impl FileOptionExtension for () {
+    fn extra_data(&self) -> Option<Arc<Vec<u8>>> {
+        None
+    }
+    fn central_extra_data(&self) -> Option<Arc<Vec<u8>>> {
+        None
+    }
+}
 /// Metadata for a file to be written
-#[derive(Clone, Debug)]
-pub struct FileOptions {
+#[derive(Clone, Debug, Copy)]
+pub struct FileOptions<T: FileOptionExtension> {
     pub(crate) compression_method: CompressionMethod,
     pub(crate) compression_level: Option<i64>,
     pub(crate) last_modified_time: DateTime,
     pub(crate) permissions: Option<u32>,
     pub(crate) large_file: bool,
     encrypt_with: Option<ZipCryptoKeys>,
-    extra_data: Arc<Vec<u8>>,
-    central_extra_data: Arc<Vec<u8>>,
+    extended_options: T,
     alignment: u16,
     #[cfg(feature = "deflate-zopfli")]
     pub(super) zopfli_buffer_size: Option<usize>,
 }
+/// Simple File Options. Can be copied and good for simple writing zip files
+pub type SimpleFileOptions = FileOptions<()>;
+/// Adds Extra Data and Central Extra Data. It does not implement copy.
+pub type FullFileOptions = FileOptions<ExtendedFileOptions>;
+/// The Extension for Extra Data and Central Extra Data
+#[derive(Clone, Debug, Default)]
+pub struct ExtendedFileOptions {
+    extra_data: Arc<Vec<u8>>,
+    central_extra_data: Arc<Vec<u8>>,
+}
+impl Sealed for ExtendedFileOptions {}
+
+impl FileOptionExtension for ExtendedFileOptions {
+    fn extra_data(&self) -> Option<Arc<Vec<u8>>> {
```

Comment:
Can these methods return the borrowed Arc instead of making a clone that will sometimes be unnecessary?

Reply:
Done